lamagraph-compiler/src/LamaML/Syntax.hs

```diff
@@ -0,0 +1,290 @@
+-}
+
+{- $lexing_blanks
+Spaces, tabs, @\\n@ and @\\r@ are ignored and are used to separate tokens.
+-}
+
+{- $lexing_comments
+Comments must be enclosed in @(*@ and @*)@.
+Nesting is allowed and must be handled.
+-}
+
+{- $lexing_idents
+@
+/ident/ ::= ( /letter/ | _ ) { /letter/ | 0...9 | _ | ' }
+
+/capitalized-ident/ ::= ( A...Z ) { /letter/ | 0...9 | _ | ' }
+
+/lowercase-ident/ ::= ( a...z | _ ) { /letter/ | 0...9 | _ | ' }
+
+/letter/ ::= A...Z | a...z
+@
+-}
+
+{- $lexing_int_lits
+@
+/integer-literal/ ::= \[-] ( 0...9 ) { 0...9 | _ }
+@
+
+/Do we need hex\/octal\/binary numbers and\/or size suffixes?/
+
+/Also no floats currently./
+-}
+
+{- $lexing_char_lits
+@
+/char-literal/ ::= ' /regular-char/ '
+               | ' /escape-sequence/ '
+
+/escape-sequence/ ::= \\ ( __|__ | " | ' | n )
+@
+
+/Do we need more escape sequences?/
+
+/We must define regular-char? Any ASCII one?/
+-}
+
+{- $lexing_string_lits
+@
+/string-literal/ ::= " { /string-character/ } "
+
+/string-character/ ::= /regular-string-character/
+                   | /escape-sequence/
+@
+
+/We must define regular-string-character? Any ASCII one?/
+
+/Do we need multiline strings?/
+-}
+
+{- $lexing_operators
+
+@
+/infix-symbol/ ::= ( = | \< | \> | @ | ^ | __|__ | & | + | - | * | / | $ | % ) { /operator-char/ }
+
+/prefix-symbol/ ::= ! { /operator-char/ }
+                | ( ? | ~ ) { /operator-char/ }+
+
+/operator-char/ ::= ! | $ | % | & | * | + | - | . | / | : | \< | = | \> | ? | @ | ^ | __|__ | ~
+@
+
+Copypasted from https://askra.de/software/ocaml-doc/4.02/lex.html#sec71, probably too complicated.
+-}
+
+{- $lexing_keywords
+Keyword table:
+
+@
+and asr begin else end false fun function
+if in land let lor lsl lsr lxor match mod
+of rec then true type when with
+
+!= && ' ( ) * + , - -> : :: ; < = > [ ]
+_ { | }
+@
+
+/I hope nothing is missing./
+-}
+
+{- $names
+/No modules currently!/
+
+@
+/value-name/ ::= /lowercase-ident/
+             | __(__ /operator-name/ __)__
+
+/operator-name/ ::= /prefix-symbol/ | /infix-op/
+
+/infix-op/ ::= /infix-symbol/
+           | * | + | - | = | != | \< | \> | || | &&
+           | mod | land | lor | lxor | lsl | lsr | asr
+
+/constr-name/ ::= /capitalized-ident/
+
+/typeconstr-name/ ::= /lowercase-ident/
+
+/field-name/ ::= /lowercase-ident/
+@
+
+/Revise or \/ || \/ lor/
+
+/Do we need records?/
+-}
+
+{- $types
+@
+/typexpr/ ::= ' /ident/
+          | _
+          | __(__ /typexpr/ __)__
+          | /typexpr/ -> /typexpr/
+          | /typexpr/ { * /typexpr/}+
+          | /typeconstr-name/
+          | /typexpr/ /typeconstr-name/
+          | __(__ /typexpr/ {, /typexpr/} __)__ /typeconstr-name/
+@
+-}
+
+{- $constants
+@
+/constant/ ::= /integer-literal/
+           | /char-literal/
+           | /string-literal/
+           | /constr-name/
+           | false
+           | true
+           | __()__
+           | begin end
```

Comment:
Wow... What is it?

Reply:
I think this is something from procedural part of OCaml and I think now that we can get rid of `begin end` constructs at all!